contracts/exchange/contracts/src/MixinAssetProxyDispatcher.sol

```diff
@@ -142,31 +155,64 @@ contract MixinAssetProxyDispatcher is
                 let dataArea := add(cdStart, 132)
+                let assetDataArea := assetData
                 // solhint-disable-next-line no-empty-blocks
                 for {} lt(dataArea, cdEnd) {} {
-                    mstore(dataArea, mload(assetData))
+                    mstore(dataArea, mload(assetDataArea))
                     dataArea := add(dataArea, 32)
-                    assetData := add(assetData, 32)
+                    assetDataArea := add(assetDataArea, 32)
                 }
 
                 /////// Call `assetProxy.transferFrom` using the constructed calldata ///////
-                let success := call(
+                didSucceed := call(
                     gas,                    // forward all gas
                     assetProxy,             // call address of asset proxy
                     0,                      // don't send any ETH
                     cdStart,                // pointer to start of input
-                    sub(cdEnd, cdStart),    // length of input  
-                    cdStart,                // write output over input
-                    512                     // reserve 512 bytes for output
+                    sub(cdEnd, cdStart),    // length of input
+                    0,                      // don't store the returndata
+                    0                       // don't store the returndata
                 )
-                if iszero(success) {
-                    revert(cdStart, returndatasize())
+
+                if iszero(didSucceed) { // Call reverted.
+                    // The asset proxies always revert with a standard string error,
+                    // which have the signature `Error(string)` and are ABI encoded
+                    // the same way as calldata for a function call is.
+                    // We will extract the string payload from this to upgrade it
+                    // to a rich revert.
+
+                    // Asset proxy revert errors have the following memory layout:
+                    // | Area            | Offset  | Length   | Contents                              |
+                    // --------------------------------------------------------------------------------
+                    // | Selector        | 0       | 4        | bytes4 function selector (0x08c379a0) |
+                    // | Params          |         |          |                                       |
+                    // |                 | 4       | 32       | uint256 offset to data (always 32)    |
+                    // | Data            |         |          |                                       |
+                    // |                 | 36      | 32       | uint256 length of data (n)            |
+                    // |                 | 68      | n        | Left-aligned message bytes            |
+
+                    // Get the size of the string data/payload.
+                    let revertDataSize := sub(returndatasize(), 36)
```

Comment:
Omitting the first 36 bytes works if we assume that the revert data is always a string, but what if we move to rich revert reasons for asset proxies too? In that case, it's probably better to just include all of the return data and decode client side based off of the function selector.

Reply:
Implemented [here](../blob/dfb43ddc18ef10e9080a579e185e8ee7b8edfb33/contracts/exchange/contracts/src/MixinAssetProxyDispatcher.sol#L182).